### ISSUE TITLE
Debugging webpack with webpack-dev-server won't work (aka when runtime scripts are not actually on disk)

### ISSUE THREAD

**Issue by conradz** [Microsoft/vscode-chrome-debug#40]
Is it possible to use this to debug JS code bundled with [webpack](http://webpack.github.io)? I tried to set it up to debug, but it wouldn't hit any breakpoints. I am using inline source maps, and got the file paths in the source maps to mirror my project file paths.

Here is how I managed to get the webpack source map paths mirroring the actual project file paths (this is just the applicable parts, obviously there is much more configuration needed for the whole project):

```
// webpack config
module.exports = {
  devtool: 'inline-source-map',
  output: {
    devtoolModuleFilenameTemplate: sourceMapName
  }
};

function sourceMapName(info) {
    var p = info.resourcePath;
    p = p.replace(/~/g, 'node_modules');
    if (path.isAbsolute(p))
        p = path.relative(root, p);
    p = path.normalize(p);
    return p.replace(/\\/g, '/');
}
```

I thought that if source maps would point to the correct file names it would work, but something doesn't seem to be working. Launching chrome and using the console works fine, but I can't set any breakpoints at all. The breakpoints show in the editor but never get hit.

**Comment by roblourens** [comment]
I see that webpack has a development server that will serve the bundle without outputting it to disk - are you using that? This extension assumes that files will be available on disk, so I don't think that will work right now. I can probably fix it but I hadn't thought a lot about that scenario.

Otherwise, I have this working with a basic webpack config. With sourcemaps, it works with the config above to fix paths, with the caveat that a breakpoint set before launching Chrome will not always bind correctly. Just filed #43 for that. But when I set the breakpoint after the page has loaded, it binds fine.

Without sourcemaps enabled, I don't see why it wouldn't work - if you set a breakpoint in the actual bundled file that is loaded in Chrome, then it's the same as any other basic scenario.

**Comment by joewood** [comment]
Yes, this is serving from **webpack-dev-server**. But aren't the JS file and map files in memory? The source files should be on disk and referenced by absolute disk file names. Is the debugger extension loading the JS files from disk?

**Comment by roblourens** [comment]
I see. Yeah the extension loads files from disk. This should be fixable.

**Comment by joewood** [comment]
That's great. Actually this explains why the "launch" option wasn't working with webpack too.
I won't bother testing this with the react-native packager. Hopefully the fix for this will work with react-native too. Debugging an app running on an iPhone in VS Code :-)

**Comment by jahewson** [comment]
I'm encountering the same problem. @roblourens, could you share your working `webpack.config.js`, `tsconfig.json`, `tasks.json`, and `launch.json`?

**Comment by roblourens** [comment]
@joewood Yeah the only difference between "launch" and "attach" is that "launch" first guesses where chrome is and starts it. Then attaches to it, and everything is the same after that. About react-native, it looks like the typical approach is to use chrome devtools - if they have a proxy that exposes the chrome remote debug protocol somehow, then it might be possible to make this work already. I'll have to look into that later.

@jahewson If you're using `webpack-dev-server` then this won't work yet. I just built a bundle.js which is really just the typical scenario.

**Comment by conradz** [comment]
OK, that seems to explain my problem as I was trying it with the webpack-dev-server. I thought that as long as the source map was inlined it wouldn't need the bundled output written to disk, but apparently that is not correct. Thanks for the quick investigation!

**Comment by joewood** [comment]
Any update on this?

**Comment by roblourens** [comment]
I'm working on this but not completely full time. I would really like to have it fixed by Wednesday. It looks like webpack has an option `--output-public-path` to actually write out scripts which might unblock you in the meantime. But there's a new webpack issue related to sourcemaps that was pointed out at the bottom of #42, so sourcemaps with webpack will be blocked on that too.

**Comment by joewood** [comment]
Thanks @roblourens appreciate it.

**Comment by roblourens** [comment]
I didn't quite finish this one - I'll keep working on it and keep you updated.

**Comment by joewood** [comment]
Great, looks like it's progressing - let me know if you want help testing or even a PR.

**Comment by roblourens** [comment]
This is more or less working in the dev branch, but needs more cleanup/testing.

**Comment by joewood** [comment]
Thanks @roblourens, did the dev branch make it in the release today, or is it still sitting in dev?

**Comment by roblourens** [comment]
It's not out yet - it turned out to be a bunch of changes and I still need to make sure I'm not breaking something else.

**Comment by Gabriel0402** [comment]
Is there any workaround like chrome's mapping network resource  to local file to attach the sources? I could make it work in the chrome debugger but still not in vscode.

**Comment by roblourens** [comment]
This should be working now if you update! Let me know how it goes.

**Comment by joewood** [comment]
Thanks @roblourens 
What kind of map files did you use in your testing? Does it need inline map files, absolute path etc?

**Comment by Gabriel0402** [comment]
Thanks @roblourens  when will the update be available?

**Comment by roblourens** [comment]
It should work with any sort of sourcemaps at this point. I tested with inline and separate files. And the update is available now, but you need to update extensions manually in Code.

**Comment by joewood** [comment]
Great, it works well. The only thing I had to change in my config was to remove the webroot setting.
Thanks for fixing this.

**Comment by skolmer** [comment]
Is it somehow possible to activate breakpoints in jsx files? I can get this to work if my files have a js file extensions but if the extension is jsx I cannot set breakpoints.

**Comment by Gabriel0402** [comment]
@roblourens  I still cannot make it work. In my launch.json I only specify the type as chrome, request as launch, url as https://localhost:3000 and sourcemap as true. And the webpack is serving the whole src file under webpack server. I still cannot set breakpoints in ts files. But when I stop during the debug mode, it'll stop at angular2 intermal module. Can you share your launch json file please? Thanks!

**Comment by joewood** [comment]
I can share mine, which works fine. I attach to chrome and generate map files from TypeScript. Files served from **public**.
```json
{
	"name": "Chrome Attach",
	"type": "chrome",
	"request": "attach",
	"sourceMaps": true,
	"port": 9222
}
```
And the typescript config:
```json
{
    "compilerOptions": {
        "outDir": "public",
        "target": "ES5",
        "module": "commonjs",
        "sourceMap": true
    }
}
```
and webpack:
```javascript
{
  entry: "./src/app.tsx",
  output: {
    filename: 'app.js',
    path: './public',
  },
  devtool:'source-map',
  devServer: {
        contentBase: "public/"
    }
}
```

**Comment by Gabriel0402** [comment]
@joewood  Thanks a lot !!

**Comment by roblourens** [comment]
@skolmer Sure, I filed #62 for that

@joewood Thanks, glad it works!

**Comment by skolmer** [comment]
@roblourens thanks!

**Comment by Gabriel0402** [comment]
@joewood  Hi, I tried to make my configuration be similar to yours, so I build my app.js and app.map.js into the public folder,  and from the log I could see that app.js is resolved and I can set breakpoints inside app.js and debug, But I got sourcemaps are enabled but app.js is a runtime script. Then the sourcemap didn't work for me then. May I ask you that did you put the sourcemap files under the same public folder? And are they point to the ts files in the private folder? And are you able to set breakpoints inside the ts files inside the private folder? Thanks! Really appreciate your response.

**Comment by jrhorn424** [comment]
We're tracking a similar issue over at https://github.com/ga-wdi-boston/js-template/issues/17. I've been doing research now and then to see if anything has changed on the webpack or Chrome side. Can anyone comment on whether our diagnosis of the issue was correct?

**Comment by ulitiy** [comment]
It still doesn't work for me.
https://gist.github.com/ulitiy/f469b392cd2d803de93cdf7121886cf3 my configs
Source maps work in the browser but not in VS Code. It attaches and pause works but breakpoints don't.

**Comment by roblourens** [comment]
Off the top of my head, you may need to set `"webRoot": "${workspaceRoot}/dist"` in your launch config instead. If that doesn't work, could you open a new issue, run it with `diagnosticLogging: true` set, and upload the log file?

**Comment by ulitiy** [comment]
I think `webRoot` makes it even worse - adds excessive dist. My webRoot is in the `workspaceRoot`
Even without webRoot:
```
  ›SourceMap: webRoot: /Users/ulitiy/Dropbox/RAILS/aurora
  ›SourceMap: no sourceRoot specified, using webRoot + script path dirname: /Users/ulitiy/Dropbox/RAILS/aurora/dist
  ›SourceMaps.scriptParsed: http://localhost:8080/dist/bundle.js was just loaded and has mapped sources: ["/Users/ulitiy/Dropbox/RAILS/aurora/dist/webpack/bootstrap 63ce91f60e62a590e8b0","/Users/ulitiy/Dropbox/RAILS/aurora/dist/src/index.ts"]
  ›SourceMaps.setBP: /Users/ulitiy/Dropbox/RAILS/aurora/src/index.ts can't be resolved to a loaded script. It may just not be loaded yet.
```
But I need it to load src from `/src`, not `/dist/src`
Can you help me with that?

**Comment by ben-x9** [comment]
Not very useful though, as I need the url option to be able to connect to the right tab

**Comment by teyc** [comment]
My webpack.config.js reads like this, and it works fine
The real trick is the `[resource-path]` section. Since my builds are output to /assets/bundle.js, I need to point the sourcemaps up one level back to the parent directory.

```
module.exports = {
  entry: './app.ts',
  output: {
    filename: 'bundle.js',
    publicPath: '/assets',
    devtoolModuleFilenameTemplate: '../[resource-path]'
  },
  resolve: {
    // Add `.ts` and `.tsx` as a resolvable extension.
    extensions: ['.webpack.js', '.web.js', '.ts', '.tsx', '.js']
  },
  devtool: 'source-map',
  devServer: {
  },
  module: {
    loaders: [
      // all files with a `.ts` or `.tsx` extension will be handled by `ts-loader`
      { test: /\.tsx?$/, loader: 'ts-loader!source-map-loader' },
      { test: /\.html$/, loader: 'raw-loader'                  }
    ]
  }
}

```

**Comment by fredrik-macrobond** [comment]
I have a pretty complex webpack-dev-server configuration, but after many tries I got it to work by removing all path and source mappings and simply having this config:

`{
            "name": "Data browser",
            "type": "chrome",
            "request": "launch",
            "url": "http://10.13.10.1:81/browser",
            "sourceMaps": true,
            "webRoot": "${workspaceRoot}"
        }`
